Adding HasCallStack to Partial functions

L0neGamer opened the pull request:
continuing the work of https://github.com/haskell/containers/pull/493/

treeowl commented on containers/src/Data/IntMap/Internal.hs:
```diff
@@ -2394,12 +2409,20 @@ lookupMin (Bin p l r) =
   | otherwise = error "findMin: empty map has no minimal element"
 
+#if __GLASGOW_HASKELL__ >= 800
+lookupMaxSure :: HasCallStack => IntMap a -> KeyValue a
```
This one is internal too. Don't add a constraint here.

treeowl commented on containers/src/Data/Sequence/Internal.hs:
```diff
@@ -2401,7 +2429,11 @@ scanl f z0 xs = z0 <| snd (mapAccumL (\ x z -> let x' = f x z in (x', x')) z0 xs
 --
 -- > scanl1 f (fromList [x1, x2, ...]) = fromList [x1, x1 `f` x2, ...]
+#if __GLASGOW_HASKELL__ >= 800
+scanl1 :: HasCallStack => (a -> a -> a) -> Seq a -> Seq a
```
I'm concerned this could build a call stack recursively. Check the Core, please.

treeowl commented on containers/src/Data/Sequence/Internal.hs:
```diff
@@ -2415,7 +2447,11 @@ scanr f z0 xs = snd (mapAccumR (\ z x -> let z' = f x z in (z', z')) z0 xs) |> z
 --
 -- __Note__: This function is partial.
+#if __GLASGOW_HASKELL__ >= 800
+scanr1 :: HasCallStack => (a -> a -> a) -> Seq a -> Seq a
```
Also concerned about recursion here.

meooow25 commented:
Hi, I need some time to review this, since I need to first understand how `HasCallstack` works and the current consensus on using them in boot libraries.

About the benchmarks, what have you tried? I would run the benchmarks for the changed functions using tasty patterns, e.g.

```
$ cabal run map-benchmarks -- -p lookup
```

(See also [CONTRIBUTING.md](https://github.com/haskell/containers/blob/master/CONTRIBUTING.md#procedure))

L0neGamer commented:
I ran `cabal bench all` before the changes, after, and then checked out master again and reran, and got very different results. 

I can certainly try running individual benchmarks; in that case I'll probably try doing "run each benchmark component individually".

treeowl commented:
@meooow25 , the constraints are used in `base`, so there shouldn't be any issue using them in other boot libraries.

treeowl commented:
FWIW, I always run `killall -STOP firefox` before running benchmarks and `killall -CONT firefox` after, but I don't claim to be good at achieving a "quiet" system.

jappeace commented:
It needs to calculate the standard deviations (also known as expected error, or volatility) to see if they're within margin per bench suite.

But I think these are kind of excessive. Nobody complained about adding callstack to unordered containers, so I doubt they will if it's added here. And if they do just revert the ones they complain about 😅

Bodigrim commented:
> Here are the percentage differences of the end results of each:

What do you mean by "end results"? Are you comparing total time as reported in "All XX tests passed"? This would make little sense.

Could you use `--csv` and `--baseline`?

meooow25 commented:
> Would the maintainers be interested if I took this opportunity to add the `x-partial` warning to all of these functions as well?

A no from me on that, sorry. I find the warnings on `head` and `tail` annoying enough.
I'm all for documenting clearly which functions are partial and the conditions under which they error. This was done in #1140, if I missed any do let me know. The rest is on the user, if they use the function they know what they are doing and there is no reason to annoy them when they compile their code.

L0neGamer commented:
As much as I would love to add `x-partial` warnings onto these functions, I don't think adding HasCallStack should be blocked on that. What do I need to do to keep this moving?

meooow25 commented on containers/src/Data/IntMap/Internal.hs:
```diff
@@ -366,6 +366,9 @@ import Text.Read
 #endif
 import qualified Control.Category as Category
 
+#if __GLASGOW_HASKELL__ >= 800
```
We support only GHC >=8.2, so all of these can simply be `#ifdef __GLASGOW_HASKELL__`

https://github.com/haskell/containers/blob/51902237d52dbb10002c5ac897330a584db8e671/containers/containers.cabal#L41-L42

treeowl commented on containers/src/Data/Sequence/Internal.hs:
```diff
@@ -516,7 +520,11 @@ instance MonadFix Seq where
 -- constant-time length and logarithmic-time indexing to speed things up.
 -- Using fromFunction, we make this about as lazy as we can.
+#if __GLASGOW_HASKELL__ >= 800
+mfixSeq :: HasCallStack => (a -> Seq a) -> Seq a
```
The reason it's not useful, I believe, is not that it's internal. The reason is that `mfix` is a class method that does not take a call stack. So there's no way to get the desired call stack info to the right place.

meooow25 commented:
This seems good overall. Benchmarks don't show anything of concern. I'm adding some more in #1182 and they seem okay too.

This will be a breaking change, so I can approve it but it will be sitting for a while until we have a major release.

L0neGamer commented:
I've fixed the `elemAt` recursion and added a changelog entry, which can be moved or adjusted as needed.

I'm not sure why this is a breaking change considering that none of the exports have changed and none of the functions should've changed in functionality.

The only potential change I can see is where the error is reported from; see below for the example change.

definitions:
<img width="466" height="117" alt="image" src="https://github.com/user-attachments/assets/df266125-6176-4851-8c16-78415b473dc2" />

usages:
<img width="774" height="429" alt="image" src="https://github.com/user-attachments/assets/4216e621-54e6-47e4-8392-0f8ba64ed4a2" />

You can see that `accessHCS` has a more useful callstack, but otherwise everything is identical.
